src/cache-metadata.ts

```diff
@@ -0,0 +1,82 @@
+import AWS from 'aws-sdk';
+import { DocumentClient } from 'aws-sdk/clients/dynamodb';
+import debug from 'debug';
+
+const log = debug('monofo:cache-metadata');
+
+const SEVEN_DAYS_SECS = 7 * 24 * 60 * 60;
+
+export interface CacheMetadataKey {
+  contentHash: string;
+  component: string;
+}
+
+export interface CacheMetadata extends CacheMetadataKey {
+  buildId: string;
+}
+
+export const CACHE_METADATA_TABLE_NAME = 'monofo_cache_metadata';
+const TableName = CACHE_METADATA_TABLE_NAME;
+
+export const CACHE_METADATA_TABLE_DEFINITION = {
+  TableName,
+  BillingMode: 'PAY_PER_REQUEST',
+  AttributeDefinitions: [
+    {
+      AttributeName: 'contentHash',
+      AttributeType: 'S',
+    },
+    {
+      AttributeName: 'component',
+      AttributeType: 'S',
+    },
+  ],
+  KeySchema: [
+    {
+      AttributeName: 'contentHash',
+      KeyType: 'HASH',
+    },
+    {
+      AttributeName: 'component',
+      KeyType: 'RANGE',
+    },
+  ],
+};
+
+export class CacheMetadataRepository {
+  private readonly client: DocumentClient;
+
+  public constructor(private readonly service: AWS.DynamoDB) {
+    this.client = new DocumentClient({ service });
+  }
+
+  public async getAll(keys: CacheMetadataKey[]): Promise<CacheMetadata[]> {
+    if (keys.length < 1) {
+      return Promise.resolve([]);
+    }
+
+    log('Getting cache metadata for keys', keys);
+    const res = await this.client
+      .batchGet({
+        RequestItems: {
+          [TableName]: {
+            Keys: keys,
+          },
+        },
+      })
+      .promise();
+
+    log('Got result', res);
+    return (res?.Responses?.[TableName] as CacheMetadata[]) || [];
+  }
+
+  public async put(item: CacheMetadata): Promise<void> {
+    log('Putting cache metadata', item);
+    await this.client
+      .put({
+        TableName,
+        Item: { ...item, expiresAt: Date.now() / 1000 + SEVEN_DAYS_SECS },
```

Comment:
Is expiry required for anything? I thought buildkite artifacts lasted forever.

Reply:
Yeah, they do. So it's not required as such.

I'm just cautious because I know if we were asked to move to our own artifact bucket, we'd want to turn on intelligent archive tiering and stuff, at _least_, but probably a lifecycle rule to remove old artifacts would be on the cards sooner rather than later. Just due to the size of some of our artifacts (that we retransmit per-build when skipping)

Wonder what our GB/day for artifacts looks like...